frontend/ui/widget/dictquicklookup.lua

```diff
@@ -807,7 +807,8 @@ function DictQuickLookup:update()
     if self.is_html then
         self.text_widget.htmlbox_widget:setContent(self.definition, self:getHtmlDictionaryCss(), Screen:scaleBySize(self.dict_font_size))
     else
-        self.text_widget.text_widget:setText(self.definition)
+        self.text_widget.text_widget.text = self.definition
+        self.text_widget.text_widget:init()
```

Comment:
I don't know if I ever tested re-using a same (Scroll)TextBoxWidget - I might feel safer re-instantiating a new one.
But if you insist, please call :free() before :init() ( as it's done in `TextBoxWidget:focus()` (which feels ugly/expensive for just a focus/unfocus, but that's another thing) - on the ScrollTextWidget if it calls it on its child TextBoxWidget, or at least on the TextBoxWidget.

Reply:
I had written that before seeing your request :)
Didn't see the initial free that would be recursive - but if there is, be sure it does free() it at this point.

Reply:
Yep, WiggetContainer is doing its job ;).

```
Highlighted Button      table: 0x743efce0
WidgetContainer: Calling free for widget        table: 0x710071f0       from    table: 0x7100e8f0
WidgetContainer: Calling free for widget        table: 0x710072b8       from    table: 0x710071f0
WidgetContainer: Calling free for widget        table: 0x7100d648       from    table: 0x710072b8
WidgetContainer: Calling free for widget        table: 0x762d7808       from    table: 0x7100d648
WidgetContainer: Calling free for widget        table: 0x762d77d0       from    table: 0x762d7808
WidgetContainer: Calling free for widget        table: 0x76230078       from    table: 0x762d77d0
WidgetContainer: Calling free for widget        table: 0x75847e38       from    table: 0x76230078
WidgetContainer: Calling free for widget        table: 0x75847d78       from    table: 0x762d7808
WidgetContainer: Calling free for widget        table: 0x75910f20       from    table: 0x75847d78
WidgetContainer: Calling free for widget        table: 0x75847da0       from    table: 0x75910f20
WidgetContainer: Calling free for widget        table: 0x7100dbb0       from    table: 0x7100d648
WidgetContainer: Calling free for widget        table: 0x75d2c278       from    table: 0x7100dbb0
WidgetContainer: Calling free for widget        table: 0x76175970       from    table: 0x75d2c278
WidgetContainer: Calling free for widget        table: 0x75905448       from    table: 0x75d2c278
WidgetContainer: Calling free for widget        table: 0x760ce460       from    table: 0x75905448
WidgetContainer: Calling free for widget        table: 0x75905510       from    table: 0x760ce460
WidgetContainer: Calling free for widget        table: 0x759054e8       from    table: 0x75905510
WidgetContainer: Calling free for widget        table: 0x75d2c240       from    table: 0x75d2c278
WidgetContainer: Calling free for widget        table: 0x7580f950       from    table: 0x75d2c240
WidgetContainer: Calling free for widget        table: 0x7100dc58       from    table: 0x7100d648
WidgetContainer: Calling free for widget        table: 0x71006ce8       from    table: 0x7100dc58
WidgetContainer: Calling free for widget        table: 0x75906940       from    table: 0x71006ce8
WidgetContainer: Calling free for widget        table: 0x710044d8       from    table: 0x75906940
WidgetContainer: Calling free for widget        table: 0x75906a08       from    table: 0x710044d8
TextBoxWidget:free      nil     on      table: 0x75906a08
WidgetContainer: Calling free for widget        table: 0x758dcc30       from    table: 0x710044d8
WidgetContainer: Calling free for widget        table: 0x7100e260       from    table: 0x7100d648
WidgetContainer: Calling free for widget        table: 0x72187ed8       from    table: 0x7100e260
WidgetContainer: Calling free for widget        table: 0x75886988       from    table: 0x72187ed8
WidgetContainer: Calling free for widget        table: 0x758ceea0       from    table: 0x75886988
WidgetContainer: Calling free for widget        table: 0x758869b0       from    table: 0x758ceea0
WidgetContainer: Calling free for widget        table: 0x75d666b0       from    table: 0x758869b0
WidgetContainer: Calling free for widget        table: 0x75d83fd8       from    table: 0x75d666b0
WidgetContainer: Calling free for widget        table: 0x762d9b28       from    table: 0x75d83fd8
WidgetContainer: Calling free for widget        table: 0x762a9828       from    table: 0x758ceea0
WidgetContainer: Calling free for widget        table: 0x758b7a28       from    table: 0x762a9828
WidgetContainer: Calling free for widget        table: 0x760cf760       from    table: 0x758b7a28
WidgetContainer: Calling free for widget        table: 0x75fcab60       from    table: 0x760cf760
WidgetContainer: Calling free for widget        table: 0x743efce0       from    table: 0x758ceea0
WidgetContainer: Calling free for widget        table: 0x743f02b8       from    table: 0x743efce0
WidgetContainer: Calling free for widget        table: 0x743f06f0       from    table: 0x743f02b8
WidgetContainer: Calling free for widget        table: 0x743efff0       from    table: 0x743f06f0
WidgetContainer: Calling free for widget        table: 0x758eb200       from    table: 0x75886988
WidgetContainer: Calling free for widget        table: 0x758eb228       from    table: 0x758eb200
WidgetContainer: Calling free for widget        table: 0x709d3e88       from    table: 0x758eb228
WidgetContainer: Calling free for widget        table: 0x74522628       from    table: 0x709d3e88
WidgetContainer: Calling free for widget        table: 0x74633b10       from    table: 0x74522628
WidgetContainer: Calling free for widget        table: 0x7586f1a0       from    table: 0x758eb200
WidgetContainer: Calling free for widget        table: 0x75d2aaf8       from    table: 0x7586f1a0
WidgetContainer: Calling free for widget        table: 0x75d58c58       from    table: 0x75d2aaf8
WidgetContainer: Calling free for widget        table: 0x75d2aa98       from    table: 0x75d58c58
WidgetContainer: Calling free for widget        table: 0x742348f8       from    table: 0x758eb200
WidgetContainer: Calling free for widget        table: 0x7575b380       from    table: 0x742348f8
WidgetContainer: Calling free for widget        table: 0x71113fd8       from    table: 0x7575b380
WidgetContainer: Calling free for widget        table: 0x7575b320       from    table: 0x71113fd8
ScrollTextWidget is     table: 0x75906940
TextBoxWidget is        table: 0x75906a08
TextBoxWidget:free      nil     on      table: 0x75906a08
```

I can leave the double-free in for documentation purposes, since it appears to be safe?

Reply:
With prettier debuginof to see what all those are:

```
Highlighted Button      table: 0x759b62d0
WidgetContainer: Calling free for widget        frontend/ui/widget/container/movablecontainer.lua       table: 0x75dabde0       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x75dabe18
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d32e0       from    frontend/ui/widget/container/movablecontainer.lua       table: 0x75dabde0
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3318       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d32e0
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x7597dce0       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3318
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x7597dca8       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x7597dce0
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x75969110       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x7597dca8
WidgetContainer: Calling free for widget        N/A     table: 0x7597d168       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x75969110
WidgetContainer: Calling free for widget        frontend/ui/widget/closebutton.lua      table: 0x75984180       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x7597dce0
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x75977bb0       from    frontend/ui/widget/closebutton.lua      table: 0x75984180
WidgetContainer: Calling free for widget        N/A     table: 0x75997d80       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x75977bb0
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x75dab9a0       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3318
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x758f2228       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x75dab9a0
WidgetContainer: Calling free for widget        N/A     table: 0x758f2200       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758f2228
WidgetContainer: Calling free for widget        frontend/ui/widget/iconbutton.lua       table: 0x7593b5e0       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758f2228
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x7593dd78       from    frontend/ui/widget/iconbutton.lua       table: 0x7593b5e0
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x759389c0       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x7593dd78
WidgetContainer: Calling free for widget        frontend/ui/widget/iconwidget.lua       table: 0x7593b680       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x759389c0
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x75914510       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758f2228
WidgetContainer: Calling free for widget        N/A     table: 0x759134e8       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x75914510
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x75daba48       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3318
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x760606e0       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x75daba48
WidgetContainer: Calling free for widget        frontend/ui/widget/scrolltextwidget.lua table: 0x763a4b60       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x760606e0
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x759a5490       from    frontend/ui/widget/scrolltextwidget.lua table: 0x763a4b60
WidgetContainer: Calling free for widget        frontend/ui/widget/textboxwidget.lua    table: 0x75df2e20       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x759a5490
TextBoxWidget:free      nil     on      table: 0x75df2e20
WidgetContainer: Calling free for widget        frontend/ui/widget/verticalscrollbar.lua        table: 0x759f7e60       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x759a5490
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d2d00       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3318
WidgetContainer: Calling free for widget        frontend/ui/widget/buttontable.lua      table: 0x758c2618       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d2d00
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3c78       from    frontend/ui/widget/buttontable.lua      table: 0x758c2618
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x758bd108       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3c78
WidgetContainer: Calling free for widget        frontend/ui/widget/button.lua   table: 0x758bd130       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758bd108
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x7588cb48       from    frontend/ui/widget/button.lua   table: 0x758bd130
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x7588cb80       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x7588cb48
WidgetContainer: Calling free for widget        N/A     table: 0x7588c5b8       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x7588cb80
WidgetContainer: Calling free for widget        frontend/ui/widget/button.lua   table: 0x710632f0       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758bd108
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x759ea3e0       from    frontend/ui/widget/button.lua   table: 0x710632f0
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x759ea418       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x759ea3e0
WidgetContainer: Calling free for widget        N/A     table: 0x758e27c0       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x759ea418
WidgetContainer: Calling free for widget        frontend/ui/widget/button.lua   table: 0x759b62d0       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758bd108
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x7580ce90       from    frontend/ui/widget/button.lua   table: 0x759b62d0
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x7580cec8       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x7580ce90
WidgetContainer: Calling free for widget        N/A     table: 0x757ce930       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x7580cec8
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x76075328       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3c78
WidgetContainer: Calling free for widget        frontend/ui/widget/button.lua   table: 0x75a86b90       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x76075328
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3e50       from    frontend/ui/widget/button.lua   table: 0x75a86b90
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3e88       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3e50
WidgetContainer: Calling free for widget        N/A     table: 0x7580dfa8       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x758d3e88
WidgetContainer: Calling free for widget        frontend/ui/widget/button.lua   table: 0x75df5928       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x76075328
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x759ee0c8       from    frontend/ui/widget/button.lua   table: 0x75df5928
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x762df5b8       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x759ee0c8
WidgetContainer: Calling free for widget        N/A     table: 0x762df310       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x762df5b8
WidgetContainer: Calling free for widget        frontend/ui/widget/button.lua   table: 0x75abcee8       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x76075328
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x75893010       from    frontend/ui/widget/button.lua   table: 0x75abcee8
WidgetContainer: Calling free for widget        frontend/ui/widget/container/widgetcontainer.lua        table: 0x75e1e680       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x75893010
WidgetContainer: Calling free for widget        N/A     table: 0x75d645a0       from    frontend/ui/widget/container/widgetcontainer.lua        table: 0x75e1e680
ScrollTextWidget is     table: 0x763a4b60
TextBoxWidget is        table: 0x75df2e20
TextBoxWidget:free      nil     on      table: 0x75df2e20
```

```diff
diff --git a/frontend/ui/widget/container/widgetcontainer.lua b/frontend/ui/widget/container/widgetcontainer.lua
index b4528bcf..990cdf14 100644
--- a/frontend/ui/widget/container/widgetcontainer.lua
+++ b/frontend/ui/widget/container/widgetcontainer.lua
@@ -113,7 +113,10 @@ end
 
 function WidgetContainer:free()
     for _, widget in ipairs(self) do
-        if widget.free then widget:free() end
+        if widget.free then
+            print("WidgetContainer: Calling free for widget", widget.init and debug.getinfo(widget.init).short_src or "N/A", widget, "from", self.init and debug.getinfo(self.init).short_src or "N/A", self)
+            widget:free()
+        end
     end
 end
 
diff --git a/frontend/ui/widget/dictquicklookup.lua b/frontend/ui/widget/dictquicklookup.lua
index e2754e85..b20b7f5e 100644
--- a/frontend/ui/widget/dictquicklookup.lua
+++ b/frontend/ui/widget/dictquicklookup.lua
@@ -774,9 +774,7 @@ function DictQuickLookup:update()
     -- Free our previous widget and subwidgets' resources (especially
     -- definitions' TextBoxWidget bb, HtmlBoxWidget bb and MuPDF instance,
     -- and scheduled image_update_action)
-    if self[1] then
-        self[1]:free()
-    end
+    self[1]:free()
 
     -- Update TextWidgets
     self.dict_title_text:setText(self.displaydictname)
@@ -802,6 +800,9 @@ function DictQuickLookup:update()
         self.text_widget.htmlbox_widget:setContent(self.definition, self:getHtmlDictionaryCss(), Screen:scaleBySize(self.dict_font_size))
     else
         self.text_widget.text_widget.text = self.definition
+        print("ScrollTextWidget is", self.text_widget)
+        print("TextBoxWidget is", self.text_widget.text_widget)
+        self.text_widget.text_widget:free()
         self.text_widget.text_widget:init()
     end
 
diff --git a/frontend/ui/widget/textboxwidget.lua b/frontend/ui/widget/textboxwidget.lua
index b5355494..b2285fab 100644
--- a/frontend/ui/widget/textboxwidget.lua
+++ b/frontend/ui/widget/textboxwidget.lua
@@ -1063,6 +1063,7 @@ function TextBoxWidget:onCloseWidget()
 end
 
 function TextBoxWidget:free(full)
+    print("TextBoxWidget:free", full, "on", self)
     -- logger.dbg("TextBoxWidget:free called")
     -- We are called with full=false from other methods here whenever
     -- :_renderText() is to be called to render a new page (when scrolling
```

Reply:
OK. But safer to keep that PR for after 2021.01 :)